src/som/GraalSOMLanguage.java

```diff
@@ -0,0 +1,104 @@
+        contextPolicy = TruffleLanguage.ContextPolicy.SHARED,
+        fileTypeDetectors = GSFileDetector.class)
+@ProvidedTags({StandardTags.CallTag.class,
+        StandardTags.StatementTag.class,
+        StandardTags.RootTag.class,
+        StandardTags.RootBodyTag.class,
+        StandardTags.ExpressionTag.class,
+        DebuggerTags.AlwaysHalt.class})
+public final class GraalSOMLanguage extends TruffleLanguage<Universe> {
+
+    public static final String ID = "GS";
+    public static final String MIME_TYPE = "application/x-graal-som";
+    public static String[] args;
+
+    @Option(help = "Help Text.", category = OptionCategory.USER) public static final OptionKey<String> TestSelector = new OptionKey<>("");
+    @Option(help = "Help Text.", category = OptionCategory.USER) public static final OptionKey<String> TestClass = new OptionKey<>("");
+    @Option(help = "Help Text.", category = OptionCategory.USER) public static final OptionKey<String> TestClasspath = new OptionKey<>("");
+
+    public GraalSOMLanguage() {
+    }
+
+    @Override
+    protected Universe createContext(final Env env) {
+        return new Universe(env, this);
+    }
+
+    @Override
+    protected void initializeContext(final Universe context) {
+        Env currentEnv = context.getEnv();
+        args = currentEnv.getApplicationArguments();
+        //TODO - on the long run, it would be better to rely on source file
+        // the arguments may not be accessed from here if this approach is used
+        // they should be pre processed in the Launcher
+//        args = context.handleArguments(args);
+//        try {
+//            context.initializeObjectSystem();
+//        } catch (ProgramDefinitionError e) {
+//            e.printStackTrace();
+//        }
+    }
+
+    @Override
+    protected OptionDescriptors getOptionDescriptors() {
+        // this class is generated by the annotation processor
+        return new GraalSOMLanguageOptionDescriptors();
+    }
+
+    @Override
+    protected boolean isObjectOfLanguage(final Object object) {
+        return object instanceof SAbstractObject;
+        //TODO - should throw an exception
+    }
+
+    public static Universe getCurrentContext() {
+        return getCurrentContext(GraalSOMLanguage.class);
+    }
+
+    @Override
+    protected CallTarget parse(final ParsingRequest request) throws Exception {
+        Universe context = getCurrentContext();
+        return Truffle.getRuntime().createCallTarget(new RootNode(this) {
+
+            @Override
+            public Object execute(final VirtualFrame frame) {
+                try {
+                    if (context.isForTesting()) {
+                        return context.interpret(args, context.testClass(), context.testSelector());
+                    }
+                    else return context.interpret(args);
+                    //TODO - leave the object system initialisation in the initializeContext method, along the lines of...
```

Comment:
Hm, interesting design question. What are the drawbacks/benefits? Is this related to the earlier comment I made? Perhaps this solution here is fine?

I guess a big question is whether we want to be able to call multiple times `eval()` on the same `engine`?

Reply:
For now, let's just support one `eval()` on the same engine and get rid of the (yet) unnecessary complexity